### ISSUE TITLE
PGD steps along the sign of the gradient

### ISSUE THREAD

**Issue by SohamTamba** [MadryLab/cifar10_challenge#18]
More of a question than an issue.

It can be infered from [here](
https://github.com/MadryLab/cifar10_challenge/blob/cd432237c2637376bfa751a254610de80e6a6d7b/pgd_attack.py#L56) the PGD steps along the sign of the gradient.

Is there any reason it does not simply step along the gradient?
i.e. `x += gradient(x)*step_size` instead of `x += sign(gradient(x))*step_size`

Thanks

**Comment by dtsip** [comment]
The idea is that you want to move each pixel by at most `step_size` in each iteration while maximizing the loss. In other words, you want to move along the gradient direction as much as possible without changing any pixel by more than `step_size`. If you think about it, this corresponds exactly to moving by `step_size * sign(grad)` (if the gradient of a pixel is positive you add `step_size`, if it is negative you subtract). In the convex optimization literature this is known as _L_infinity_-based gradient descent.

In principle, you could also just take steps along the gradient. However, we found that this takes much longer to converge and makes tuning the learning rate harder.
